### ISSUE TITLE
Docker hangs when using multiple hosts 

### ISSUE THREAD

**Issue by vilmara** [uber/horovod#272]
Hi all, I have 2 hosts with 4 GPUs each one, I am trying to run the test inside the docker. I  was able to set up passwordless authentication as a root user, the host where mpirun is executed is able to ssh to the secondary host without any prompts. Then, when I run the below commands on each host they hang forever without any output, here are the commands I am using.

Primary worker:
root@host-1:~# nvidia-docker run -it --network=host -v /root/.ssh:/root/.ssh horovod:latest
root@host-1:/examples# mpirun -np 8 \ 
-H 192.168.10.1:4,192.168.10.2:4 \ 
-bind-to none -map-by slot \ 
-x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
-x NCCL_SOCKET_IFNAME=^docker0 \
-mca pml ob1 -mca btl ^openib \
-mca btl_tcp_if_exclude lo,docker0 \ 
-mca plm_rsh_args "-p 12345" -vvvv 192.168.10.2 \
python keras_mnist_advanced.py

Secondary worker:
root@host-2:~# nvidia-docker run -it --network=host -v /root/.ssh:/root/.ssh horovod:latest bash -c "/usr/sbin/sshd -p 12345; sleep infinity"

some recommendations please?

**Comment by alsrgv** [comment]
Are you sure your first command is copy-pasted correctly?

`-vvvv` should be inside `plm_rsh_args` and `192.168.10.2` should only be in the host list.

Assuming those are fixed, are you able to ssh to second host container from first host container?

I.e. `ssh -p 12345 root@192.168.10.2` from within the first container.

**Comment by vilmara** [comment]
Hi @alsrgv , I followed the examples in [#226](https://github.com/uber/horovod/issues/226) and was able to run the test keras_mnist_advanced.py as well as the TensorforFlow benchmarks with real data imagenet. In both cases it produced expected metrics but also the warning "WARN Failed to open libibverbs.so", does it affect the final performance of the model?, what does it mean?, what is the problem?

...a lot of outputs...
C4130-1:25:232 [0] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-1:25:232 [0] INFO Using internal Network Socket
C4130-1:25:232 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
C4130-1:25:232 [0] INFO NET : Using interface ib0:192.168.10.1<0>
C4130-1:25:232 [0] INFO NET/Socket : 1 interfaces found
NCCL version 2.1.15+cuda9.0

C4130-1:27:229 [2] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-1:27:229 [2] INFO Using internal Network Socket
C4130-1:27:229 [2] INFO Using NCCL Low-latency algorithm for sizes below 16384

C4130-1:26:234 [1] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-1:26:234 [1] INFO Using internal Network Socket
C4130-1:26:234 [1] INFO Using NCCL Low-latency algorithm for sizes below 16384

C4130-1:28:238 [3] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-1:28:238 [3] INFO Using internal Network Socket
C4130-1:28:238 [3] INFO Using NCCL Low-latency algorithm for sizes below 16384

C4130-2:24:237 [0] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-2:24:237 [0] INFO Using internal Network Socket
C4130-2:24:237 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384

C4130-2:26:228 [2] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-2:26:228 [2] INFO Using internal Network Socket
C4130-2:26:228 [2] INFO Using NCCL Low-latency algorithm for sizes below 16384

C4130-2:27:234 [3] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
C4130-2:27:234 [3] INFO Using internal Network Socket
C4130-2:27:234 [3] INFO Using NCCL Low-latency algorithm for sizes below 16384

C4130-2:25:229 [1] misc/ibvwrap.cu:60 WARN Failed to open libibverbs.so[.1]
... a lot of outpouts...

final output with the test keras_mnist_advanced.py
INFO:tensorflow:loss = 0.0, step = 2490 (0.197 sec)
INFO:tensorflow:Saving checkpoints for 2500 into ./checkpoints/model.ckpt.

final output with the test tf_cnn_benchmarks.py 
----------------------------------------------------------------
100     images/sec: 157.5 +/- 1.3 (jitter = 11.2)       8.041
----------------------------------------------------------------
total images/sec: 1259.63
----------------------------------------------------------------

Finally, where can I find benchmarks results with Hovored using different nvidia gpu models to compare these with my results?

Thanks,

Vilmara

**Comment by alsrgv** [comment]
Assuming your cluster does not have RoCE or InfiniBand NICs, you can ignore `WARN Failed to open libibverbs.so` message - it simply means that NCCL cannot use RDMA.

You can look at [this page](https://github.com/uber/horovod/blob/master/docs/benchmarks.md) for benchmark results and instructions how to reproduce those on Pascal GPUs.

**Comment by vilmara** [comment]
Hi @alsrgv, the clusters have mellanox cards for the Infiniband network, but I am still getting the same warnings, how can I fix it?, does it will affect the performance?

**Comment by alsrgv** [comment]
@vilmara, I don't think either of those has any adverse performance effects.

**Comment by vilmara** [comment]
@alsrgv, how can I make the additions explained in [Usage](https://github.com/uber/horovod#usage) from the command line?. do you have some example?. I am trying to save the checkpoints passing the flag --data_dir to the tensorflow benchmarks, but getting the below error:

< a lot of normal outputs>
....
2018-05-23 14:23:47.184453: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at save_restore_v2_ops.cc:184 : Not found: Key v0/cg/affine0/biases not found in checkpoint
I0523 14:23:47.192735 140012935317248 tf_logging.py:116] Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.NotFoundError'>, Key v0/cg/affine0/biases not found in checkpoint
         [[Node: save/RestoreV2 = RestoreV2[dtypes=[DT_INT64, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]
...
<a lot of error messages>
...
NotFoundError (see above for traceback): Key v0/cg/affine0/biases not found in checkpoint
         [[Node: save/RestoreV2 = RestoreV2[dtypes=[DT_INT64, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

**Comment by alsrgv** [comment]
@vilmara, if you're running TF benchmarks, you can just pass `--variable_update horovod`, but if you're writing your own model you'll have to make those changes in your code.

As for your restore problem, is it possible that there is some old checkpoint?
